pkg/dockerfile/generator.go

```diff
@@ -131,9 +132,53 @@ func (g *Generator) baseImage() (string, error) {
 }
 
 func (g *Generator) preamble() string {
-	return `ENV DEBIAN_FRONTEND=noninteractive
-ENV PYTHONUNBUFFERED=1
-ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/lib/x86_64-linux-gnu:/usr/local/nvidia/lib64:/usr/local/nvidia/bin`
+	// If given, variables should be a list of strings, formatted like `KEY=VALUE`.
+	environmentVariables := g.Config.Build.BuildEnv
+
+	// Regex for valid environment variable names
+	regexpVariableName := regexp.MustCompile("^[A-Za-z_][A-Za-z0-9_]*$")
+
+	// Parse them into a list of strings. We also preserve the order.
+	// TODO(hangtwenty): Clean up by using an ordered map (maybe from a library)
+	envVarKeys := []string{}
+	envVarMap := make(map[string]string)
+	if len(environmentVariables) > 0 {
+		envVarMap = make(map[string]string)
+		for _, v := range environmentVariables {
+			parts := strings.SplitN(v, "=", 2)
+
+			if ok := regexpVariableName.MatchString(parts[0]) && len(parts) == 2; ok {
+				envVarMap[parts[0]] = parts[1]
+				envVarKeys = append(envVarKeys, parts[0])
+			}
+		}
+	}
+
+	if _, ok := envVarMap["XDG_CACHE_HOME"]; !ok {
+		// Cog sets a default value for $XDG_CACHE_HOME. Why:
+		// Cog mounts the project directory so anything in there (in /src in the
+		// image) gets retained between runs. $XDG_CACHE_HOME is used by various
+		// libraries including popular ML libraries; so, setting it to a subdirectory
+		// within /src results in retaining the cache between runs. Ultimately,
+		// everything in /src gets "baked in" to the cog image. So this is great
+		// for caching pretrained models and so on.
+		// For more context see: https://github.com/replicate/cog/issues/320
+		envVarMap["XDG_CACHE_HOME"] = "/src/.cache"
+	}
+	if !sliceContains(envVarKeys, "XDG_CACHE_HOME") {
+		envVarKeys = append([]string{"XDG_CACHE_HOME"}, envVarKeys...)
+	}
+
+	var envVarLines []string
+	for _, envVarName := range envVarKeys {
+		v := envVarMap[envVarName]
+		envVarLines = append(envVarLines, fmt.Sprintf("ENV %s=%s", envVarName, v))
+	}
+
+	return `ENV DEBIAN_FRONTEND=noninteractive \
+	PYTHONUNBUFFERED=1 \
```

Comment:
- [ ] I should remove the `\` changes here; I think it would cause a layer to rebuild, even for projects that do not adopt `build.env`. Instead, lead the existing `ENV DEBIAN_FRONTEND=...`/`ENV PYTHONUNBUFFERED=`... as separate lines. (In other words, just avoiding unnecessarily invalidating cached layers for models re-built after this change.)